pkg/eestream/encode.go

```diff
@@ -111,16 +143,19 @@ type encodedPiece struct {
 }
 
 func (ep *encodedPiece) Read(p []byte) (n int, err error) {
-	// lock! threadsafety matters here
-	ep.er.cv.L.Lock()
-	defer ep.er.cv.L.Unlock()
-
+	if ep.er.err != nil {
+		return 0, ep.er.err
+	}
 	outbufs, i := ep.er.outbufs, ep.i
 	if len(outbufs[i]) <= 0 {
-		// if we don't have any buffered result yet, wait until we do
-		err := ep.er.wait()
-		if err != nil {
-			return 0, err
+		// take the next block from the cannel or block if channel is empty
+		var ok bool
+		outbufs[i], ok = <-ep.er.chans[i]
+		if !ok {
+			// channel is closed
+			// TODO should be different error if channel closed for slowliness
+			// which is better: io.ErrUnexpectedEOF or io.ErrClosedPipe?
```

Comment:
my opinion is `ErrUnexpectedEOF`

Reply:
reason being ```If the EOF occurs unexpectedly in a structured data stream, the appropriate error is either ErrUnexpectedEOF or some other error giving more detail. ``` and I don't see an error that necessarily provides more detail

Reply:
Thanks for the advice. I will follow it.